rust/arrow/src/datatypes.rs

```diff
@@ -1142,6 +1142,44 @@ impl DataType {
                 | Float64
         )
     }
+
+    /// Compares this data type with another data type only based on the data type
+    /// including nested data types, but not based on other values.
+    pub fn cmp_type(&self, other: &Self) -> bool {
+        match (self, other) {
+            (DataType::List(f1), DataType::List(f2)) => {
+                f1.data_type().cmp_type(f2.data_type())
+            }
+            (DataType::FixedSizeList(f1, _), DataType::FixedSizeList(f2, _)) => {
+                f1.data_type().cmp_type(f2.data_type())
+            }
+            (DataType::LargeList(f1), DataType::LargeList(f2)) => {
+                f1.data_type().cmp_type(f2.data_type())
+            }
+            (DataType::Struct(f1), DataType::Struct(f2)) => {
+                if f1.len() == f2.len() {
```

Comment:
Can all use the form `f1.len() == f2.len() && f1.iter()...` without if else

Reply:
Yeah that looks much better.